src/CoreWCF.Primitives/src/CoreWCF/Description/DispatcherBuilder.cs

```diff
@@ -599,9 +600,25 @@ private static void BuildProxyOperation(OperationDescription operation, ClientRu
             {
                 child.AuthorizationPolicy = new Lazy<AuthorizationPolicy>(() =>
                 {
+                    object serviceInstance = OperationContext.Current.InstanceContext.GetServiceInstance();
```

Comment:
This code has a few unintended consequences for some scenarios.

As you've noticed, the call to InstanceBehavior.EnsureInstanceContextAsync happens after the call to authorizationBehavior.AuthorizeAsync so you moved it earlier. There's another call that needs to also happen earlier. Take a look at InstanceBehavior.EnsureServiceInstance. It has an optional capability of releasing an existing service instance before creating a new one. If that's the case, you could double allocate with this change. If some work is done in the constructor of the instance, this could have some bad side effects. We need to use the call to EnsureServiceInstance to allocate the instance and only call it once. InstanceContext.GetInstance will then return a cached value so won't double allocate. This would fix the next issue I discovered which I'm leaving here to increase your understanding but will be resolved by fixing this.  

If a custom IInstanceProvider is being used, this results in GetInstance being called without the request message. The built-in implementation won't have a change in behavior, but a custom implementation might. We need to pass the request message, which you can get from  OperationContext.RequestContext.RequestMessage. InstanceContext.GetServiceInstance caches the instance so if we use EnsureServiceInstance before this code, this becomes a non-issue as it will already have been done correctly.   

I was initially thinking we should optionally call EnsureInstanceContextAsync and EnsureServiceInstance earlier when enabling this feature but then I took a look at what else happens between authorizationBehavior.AuthorizeAsync and EnsureServiceInstance and there are just too many potential changes in behavior. One example is we allocate the service instance while running within a captured synchronization context. This could be important if running in a UI app like WPF of WinForms and the instance constructor accesses UI elements. It also happens after any concurrency locks are acquired. This means we could now allocate more instances than intended which could have a knock on effect elsewhere (more connections to backend systems for example). I've been working with a customer on WCF recently that was using some of these extensibility points which moving the instance creation earlier would have broken badly.

I now think we need to invert the thinking and instead, only if policies are being used move the authorization to later in the pipeline after the call to EnsureServiceInstance. If policies aren't being used, then we leave it where it is.  

This will need to be mentioned in your blog post. There are some things which will now happen before authorization which didn't before. 
1. If using a dynamic quota throttle, it will be acquired before authorization
2. IDispatchMessageInspector.AfterReceiveRequest will be called
3. Concurrency lock will be acquired
4. Authorization will now run on a captured SynchronizationContext

I don't think this is a big problem as the old Windows Permissions attribute for requiring the caller to be in a particular Windows domain group for example was done by the CLR just before calling the operation method which is as late as you can get and that was an authorization mechanism.  

Can you open an issue to move the authorization step to permanently be later for the 2.0 version with an AppContext switch to revert the behavior. The plan would be to remove the AppContext switch in 3.0 if it doesn't cause problems.